### DISCUSSION TITLE
Presets as file in same machine

### DISCUSSION THREAD

**Discussion by rubasace** [renovatebot/renovate#8299] [Request Help] [answered]
I'm running renovate as a k8s cronjob, connected to Bitbucket Cloud. 

I know there's no integration with Bitbucket Cloud for presets, so I'm currently storing my configuration on a Github repository. All working fine till here. 

I was wondering if there's any way to refer to presets on the same machine. I would definitely prefer to keep my presets as configmaps and mount them to the container, instead of referring to any repo (even if it were Bitbucket).

**Comment by viceice** [comment]
That's currently not supported. check the [docs](https://docs.renovatebot.com/config-presets/) for supported shared preset hosting

**Reply by viceice** [reply]
#7448, #8217

**Reply by rubasace** [reply]
sorry, I don't see any mention about files from the same machine in the above-mentioned issues. Not sure I've explained myself properly, I would like to be able to load files from the container directories (i.e. `/opt/renovate/extra-presets/mypreset.json`), not from any external source.

**Reply by rarkins** [reply]
In that case it would seem to make more sense that you use the bot's `config.js` to import the presets into its config instead of extending them.

**Reply by rubasace** [reply]
that sounds very interesting, is there any way to do it with a json configuration? or would I need to use the JS one to be able to import extra files?. Also, is there any example of this available?

**Reply by rarkins** [reply]
You can use `config.json` instead, but a JSON file won't let you dynamically include from other files at runtime the same way a JS file would.

**Reply by rubasace** [reply]
is there any example you can show me or any link to somewhere this is explained, as I'm not familiar with js configurations at all?

**Reply by rarkins** [reply]
```js
module.exports = {
 // insert any valid JS code here such that the result is a valid Renovate configuration object
}
```

**Reply by rubasace** [reply]
got it. So you are suggesting something like this:

```js
  const myPreset = require('./myPreset.js');

  module.exports = {
    onboardingConfig: {
      extends: [ 'config:base'],
      ...myPreset
    }
  };
```

The problem with this approach is that it goes against the idea of the presets: it's true that the onboarding will be configured properly, but if I change the preset after that, all already onboarded services won't reflect the change. 

I honestly think having a preset in a local file would be a nice feature, as it's also compatible with all git repositories and avoids extra configuration for teams like mine, where we have all configurations as k8s configmaps.

What do you think?

**Reply by viceice** [reply]
I think a faster solution is to implement local preset for bitbucket cloud.

Checkout the bitbucket Server implementation. Should be easy to adapt for cloud.

https://github.com/renovatebot/renovate/blob/master/lib/config/presets/bitbucket-server/index.ts

**Reply by viceice** [reply]
@rarkins a like the idea of `file>dir/file` preset's. Renovate would read `./dir/file.json`. 🤔

**Reply by rarkins** [reply]
I don't object to it, but I don't see much to gain by it

**Reply by rubasace** [reply]
From my point of view, there are a few, but they may not be relevant to others:

1. More flexibility in terms of being able to provide the local file as you want, which might adapt better to certain teams workflows. For example, as ConfigMaps or even as secrets on k8s that can be generated from whatever git provider as part of a CI/CD pipeline.
2. Day-zero support of presets on any potential brand new git integration, without forcing to go cross-platform (as it's the current state with Bitbucket Cloud).

But as I said, those points are relevant to me but I've been using your product for roughly a week now. You definitely have a clearer picture of what's relevant here, I'm just trying to contribute my two cents.

**Comment by rubasace** [comment, marked as answer]
Hi guys, I know it's been a long time but would it be possible to give another look at this? 
Currently, we are starting to rollout Renovate to the entire company and our security team isn't happy with us having the usernames exposed in a public Github repo. 
I'd personally still like to be able to keep our presets in the same repo, mounting them from a k8s configmap, for that I'd need to be able to import presets from a disk location with whatever syntax you find best.

**Reply by rarkins** [reply]
Presets should be supported in Bitbucket now

**Reply by rubasace** [reply]
Is there anywhere in the documentation where I can read about how to set it up? 
I couldn't find a single mention of Bitbucket on the [presets documentation](https://docs.renovatebot.com/config-presets)

PS: We are using Bitbucket Cloud

**Reply by rarkins** [reply]
https://docs.renovatebot.com/config-presets/#local-presets

**Reply by rubasace** [reply]
I got them working, thanks for that!

Now I'm trying the `migratePresets` part, so I don't have to change all repos by hand. But I'm struggling: My `config.json` contains this:

```js
module.exports = {
  platform: 'bitbucket',
  migratePresets: {
    'github>mycompany/preset:common': 'local>mycompany/renovate-cronjob//presets/common',
    'github>mycompany/preset:preset1': 'local>mycompany/renovate-cronjob//presets/preset1',
    'github>mycompany/preset:preset2': 'local>mycompany/renovate-cronjob//presets/preset2',
},
...
```

But Renovate isn't migrating any of the onboarded repositories presets... Am I missing something? As of documentation it seems to be ok but couldn't find many examples

**Reply by rarkins** [reply]
What result do you expect to see and what are you seeing instead?

**Reply by rubasace** [reply]
I was expecting Renovate to either create PRs to update the `renovate.json` files on the onboarded repos or to directly do the replacements in the code, but I'm not seeing any of those. Nothing that I can identify in the logs

**Reply by rarkins** [reply]
Migration by default is done internally, meaning you don't need to make the changes to each repo. You can try this new option if you want a PR: https://docs.renovatebot.com/configuration-options/#configmigration (maybe try one repo at a time to begin with instead of enabling it globally)

**Reply by rubasace** [reply]
Yeah!! it worked!! Thanks a lot :)
